<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of getPhylDist</title>
  <meta name="keywords" content="getPhylDist">
  <meta name="description" content="getPhylDist">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../index.html">Home</a> &gt;  <a href="index.html">RAVEN</a> &gt; getPhylDist.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../index.html"><img alt="<" border="0" src="../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for RAVEN&nbsp;<img alt=">" border="0" src="../right.png"></a></td></tr></table>-->

<h1>getPhylDist
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>getPhylDist</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>function phylDistStruct=getPhylDist(keggPath,onlyInKingdom) </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre class="comment"> getPhylDist
   Calculates distance between species in KEGG based on systematic name.

   keggPath        pathway to the KEGG database files
   onlyInKingdom   if true, it generates a distance matrix with distance 
                   Inf for organisms from another domains
                   (Prokaryota, Eukaryota) (opt, default false)

   phylDistStruct  a structure with a list of organism ids and a matrix
                   that specifies their pairwise distances

   This simple metric is based on the number of nodes two organisms are
   away from each other in KEGG.

   Usage: phylDistStruct=getPhylDist(keggPath,onlyInKingdom)

   Rasmus Agren, 2012-12-17</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../matlabicon.gif)">
<li><a href="downloadKEGG.html" class="code" title="function exitFlag=downloadKEGG(keggPath)">downloadKEGG</a>	downloadKEGG</li></ul>
This function is called by:
<ul style="list-style-image:url(../matlabicon.gif)">
<li><a href="getKEGGModelForOrganism.html" class="code" title="function model=getKEGGModelForOrganism(organismID,fastaFile,dataDir,outDir,keepUndefinedStoich,keepIncomplete,keepGeneral,cutOff,minScoreRatioG,minScoreRatioKO,maxPhylDist,nSequences)">getKEGGModelForOrganism</a>	getKEGGModelForOrganism</li></ul>
<!-- crossreference -->



<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre>0001 <a name="_sub0" href="#_subfunctions" class="code">function phylDistStruct=getPhylDist(keggPath,onlyInKingdom)</a>
0002 <span class="comment">% getPhylDist</span>
0003 <span class="comment">%   Calculates distance between species in KEGG based on systematic name.</span>
0004 <span class="comment">%</span>
0005 <span class="comment">%   keggPath        pathway to the KEGG database files</span>
0006 <span class="comment">%   onlyInKingdom   if true, it generates a distance matrix with distance</span>
0007 <span class="comment">%                   Inf for organisms from another domains</span>
0008 <span class="comment">%                   (Prokaryota, Eukaryota) (opt, default false)</span>
0009 <span class="comment">%</span>
0010 <span class="comment">%   phylDistStruct  a structure with a list of organism ids and a matrix</span>
0011 <span class="comment">%                   that specifies their pairwise distances</span>
0012 <span class="comment">%</span>
0013 <span class="comment">%   This simple metric is based on the number of nodes two organisms are</span>
0014 <span class="comment">%   away from each other in KEGG.</span>
0015 <span class="comment">%</span>
0016 <span class="comment">%   Usage: phylDistStruct=getPhylDist(keggPath,onlyInKingdom)</span>
0017 <span class="comment">%</span>
0018 <span class="comment">%   Rasmus Agren, 2012-12-17</span>
0019 <span class="comment">%</span>
0020 
0021 <span class="keyword">if</span> nargin&lt;2
0022     onlyInKingdom=false;
0023 <span class="keyword">end</span>
0024 
0025 <span class="comment">%Check if the reactions have been parsed before and saved. If so, load the</span>
0026 <span class="comment">%model.</span>
0027 [ST I]=dbstack(<span class="string">'-completenames'</span>);
0028 ravenPath=fileparts(ST(I).file);
0029 distFile=fullfile(ravenPath,<span class="string">'kegg'</span>,<span class="string">'keggPhylDist.mat'</span>);
0030 <span class="keyword">if</span> exist(distFile, <span class="string">'file'</span>)
0031     fprintf([<span class="string">'NOTE: Importing KEGG phylogenetic distance matrix from '</span> strrep(distFile,<span class="string">'\'</span>,<span class="string">'/'</span>) <span class="string">'.\n'</span>]);
0032     load(distFile);
0033 <span class="keyword">else</span>
0034     <span class="comment">%Download required files from KEGG if it doesn't exist in the directory</span>
0035     <a href="downloadKEGG.html" class="code" title="function exitFlag=downloadKEGG(keggPath)">downloadKEGG</a>(keggPath);
0036 
0037     <span class="comment">%Open the file that describes the naming of the species</span>
0038     fid = fopen(fullfile(keggPath,<span class="string">'taxonomy'</span>), <span class="string">'r'</span>);
0039 
0040     phylDistStruct.ids={};
0041 
0042     <span class="comment">%Keeps the categories for each organism</span>
0043     orgCat={};
0044 
0045     currentCat={}; <span class="comment">%Keeps track of the current category</span>
0046     depth=0; <span class="comment">%Keeps track of the current tree depth</span>
0047 
0048     <span class="comment">%Loop through the file</span>
0049     orgCounter=0;
0050     <span class="keyword">while</span> 1
0051       <span class="comment">%Get the next line</span>
0052       tline = fgetl(fid);
0053 
0054       <span class="comment">%Abort at end of file</span>
0055       <span class="keyword">if</span> ~ischar(tline)
0056           <span class="keyword">break</span>;
0057       <span class="keyword">end</span>
0058 
0059       <span class="keyword">if</span> any(tline)
0060           <span class="comment">%Check if it's a new category</span>
0061           <span class="keyword">if</span> tline(1)==<span class="string">'#'</span>
0062               <span class="comment">%Find the first space (=depth +1)</span>
0063               sPos=strfind(tline,<span class="string">' '</span>)-1; <span class="comment">%Should always exist</span>
0064               sPos=sPos(1);
0065 
0066               <span class="comment">%If we have stepped back one step in the tree</span>
0067               <span class="keyword">if</span> sPos&lt;depth
0068                   currentCat=currentCat(1:sPos);
0069               <span class="keyword">end</span>
0070               depth=sPos;
0071 
0072               currentCat{depth}=tline(sPos+2:end);
0073           <span class="keyword">else</span>
0074               orgCounter=orgCounter+1;
0075               <span class="comment">%It is an organism</span>
0076               <span class="comment">%Get the id between first and second white space</span>
0077               sPos=find(isstrprop(tline, <span class="string">'wspace'</span>)); <span class="comment">%Should always exist</span>
0078               phylDistStruct.ids{orgCounter}=tline(sPos(1)+1:sPos(2)-1);
0079               orgCat{orgCounter}=currentCat;
0080           <span class="keyword">end</span>
0081       <span class="keyword">end</span>
0082     <span class="keyword">end</span>
0083 
0084     <span class="comment">%Generate a distance matrix (very straight forward here, not neat)</span>
0085     phylDistStruct.distMat=zeros(numel(phylDistStruct.ids));
0086     <span class="keyword">for</span> i=1:numel(phylDistStruct.ids)
0087         <span class="keyword">for</span> j=1:numel(phylDistStruct.ids)
0088             <span class="keyword">if</span> onlyInKingdom==true
0089                <span class="keyword">if</span> ~strcmp(orgCat{i}(1),orgCat{j}(1))
0090                    phylDistStruct.distMat(i,j)=Inf;
0091                    <span class="keyword">continue</span>;
0092                <span class="keyword">end</span>
0093             <span class="keyword">end</span>
0094             <span class="comment">%Calculate the distance between then</span>
0095             dist=numel(orgCat{i})-numel(orgCat{j});
0096             <span class="keyword">if</span> dist&gt;0
0097                 aCat=orgCat{i}(1:end-dist);
0098             <span class="keyword">else</span>
0099                 aCat=orgCat{i};
0100             <span class="keyword">end</span>
0101             <span class="keyword">if</span> dist&lt;0
0102                 bCat=orgCat{j}(1:end+dist);
0103             <span class="keyword">else</span>
0104                 bCat=orgCat{j};
0105             <span class="keyword">end</span>
0106 
0107             <span class="comment">%Loop through the categories and stop when they are the same</span>
0108             <span class="keyword">for</span> k=numel(aCat):-1:1
0109                <span class="keyword">if</span> strcmp(aCat{k},bCat{k})
0110                   <span class="keyword">break</span>; 
0111                <span class="keyword">end</span>
0112             <span class="keyword">end</span>
0113             phylDistStruct.distMat(i,j)=dist+numel(aCat)-k;
0114         <span class="keyword">end</span>
0115     <span class="keyword">end</span>
0116     <span class="comment">%Save the structure</span>
0117     save(distFile,<span class="string">'phylDistStruct'</span>);
0118 <span class="keyword">end</span>
0119 <span class="keyword">end</span></pre></div>
<hr><address>Generated on Mon 06-Jan-2014 14:58:12 by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>